### DISCUSSION TITLE
guides/yolo-data-augmentation/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#21177] [Docs]
# guides/yolo-data-augmentation/

Learn about essential data augmentation techniques in Ultralytics YOLO. Explore various transformations, their impacts, and how to implement them effectively for improved model performance.

https://docs.ultralytics.com/guides/yolo-data-augmentation/

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for bringing up Ultralytics YOLO data augmentation techniques 🚀! For new users, we recommend checking out the [Docs](https://docs.ultralytics.com/) where you'll find helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as well as answers to many common questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so our team can assist you more efficiently.

If you have a custom training ❓ Question, please share as much context as possible, such as dataset image samples and training logs, and ensure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Connect with the Ultralytics community:
- For chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- Or participate in our [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

To ensure you have the latest features and bug fixes, upgrade `ultralytics` and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs in the following verified, up-to-date environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help guide you. An Ultralytics engineer will review and assist you soon! 😊

**Comment by dolokbutarbutar24** [comment]
is that randaugment choosing randomly augmentation like shear or other augmentation technique? i dont understand the differences between randaugment and normal augment like shear and other

**Reply by dolokbutarbutar24** [reply]
is it woth to use if my data is plant disease image that sensitive to local feature?

**Reply by glenn-jocher** [reply]
`randaugment` is **not** “randomly choosing `shear`/`mosaic`/etc.” — in Ultralytics it’s the `auto_augment='randaugment'` option used **only for `classify`**, and it randomly samples from a predefined torchvision-style augmentation policy; the “normal” YOLO args like `shear`, `degrees`, `mosaic`, `mixup`, `cutmix` are separate controls used in the `detect/segment/pose/obb` augmentation pipeline (see the [Data Augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/) and the [augmentation arguments table](https://docs.ultralytics.com/usage/cfg/)).  

For plant disease images (often very local/texture-sensitive), it can be worth trying, but start conservative and validate: if RandAugment hurts fine lesion detail, disable it (and `erasing`) and rely on mild color jitter/flip instead, e.g.  
```python
from ultralytics import YOLO

model = YOLO("yolo11n-cls.pt")
model.train(data="your_dataset", auto_augment=None, erasing=0.0)
```

**Comment by shkim-lp** [comment]
Hello, I have 2 questions with respect to hsv_s, the saturation adjustment.

1. Google AI claims "hsv_s = 1.0" results in adjustment of input images' saturation level to any number within (0,1) range.
Since there is an infinite number of different real values within (0.0, 1.0) range, I reckon that larger number of epochs entails more diverse saturation levels. Am I correct?

2. Every image within the same batch can get assigned a different hsv_s value? 

Thank you for your clarification!

**Reply by glenn-jocher** [reply]
`hsv_s` in Ultralytics YOLO is a per-image random jitter magnitude: on each time an image is augmented, a (continuous) random saturation adjustment is sampled within the configured range, so more training iterations/epochs generally expose the model to more saturation variations, but repeats are normal since it’s stochastic (it’s not a finite set). Yes, it’s sampled independently per image, so different images in the same batch can receive different saturation adjustments; see the [YOLO data augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/) and the [RandomHSV reference](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.RandomHSV).
